File: java/剑指offer/T30.java
package 剑指offer;

/**
 * 定义栈的数据结构，请在该类型中实现一个能够得到栈的最小元素的 min 函数在该栈中，调用 min、push 及 pop 的时间复杂度都是 O(1)。
 *
 * 示例:
 *
 * MinStack minStack = new MinStack();
 * minStack.push(-2);
 * minStack.push(0);
 * minStack.push(-3);
 * minStack.min();   --> 返回 -3.
 * minStack.pop();
 * minStack.top();      --> 返回 0.
 * minStack.min();   --> 返回 -2.
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/bao-han-minhan-shu-de-zhan-lcof
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.min();
 */

public class T30 {
    /** initialize your data structure here. */
    Stack<Integer> A,B;
    public T30() {
        A=new Stack<>();
        B=new Stack<>();
    }

    public void push(int x) {
        A.push(x);
        if (B.empty()||B.peek()>=x){
            B.push(x);
        }
    }

    public void pop() {
        int top=A.pop();
        if (top==B.peek()){
            B.pop();
        }
    }

    public int top() {
        return A.peek();
    }

    public int min() {
        return B.peek();
    }
}

File: rust/src/offer/t30.rs
struct MinStack {
    A: Vec<i32>,
    B: Vec<i32>,
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(x);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.min();
 */
/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {
    /** initialize your data structure here. */
    fn new() -> Self {
        Self {
            A: vec![],
            B: vec![],
        }
    }

    fn push(&mut self, x: i32) {
        self.A.push(x);
        match self.B.last() {
            Some(&last) => if last >= x { self.B.push(x) },
            None => self.B.push(x),
        };
    }

    fn pop(&mut self) {
        let pop=self.A.pop().unwrap();
        if pop==*self.B.last().unwrap() {
            self.B.pop().unwrap();
        }
    }

    fn top(&self) -> i32 {
        *self.A.last().unwrap()
    }

    fn min(&self) -> i32 {
        *self.B.last().unwrap()
    }
}

